File: imperativo/livrodb.cpp
#include <iostream>
#include <sqlite3.h>
#include <vector>

using namespace std;

void criaTabelaLivro();
int editaLivro(struct Livro &livro);
int insereLivro(struct Livro &livro);
int geraIDLivro();

int listaLivros(vector<struct Livro> &livros);

int recomendaLivros(vector<struct Livro> &livros, struct Usuario &usuario);

int removeLivro(int id);

void criaTabelaLivro() {
    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao criar a tabela de livro: ";

    if (retorno) {
        cerr << "Não foi possível abrir o banco de dados: " << sqlite3_errmsg(bancoDados) << endl;
    }

    string sql = "CREATE TABLE IF NOT EXISTS livro("
                 "id INT PRIMARY KEY NOT NULL, "
                 "nome TEXT NOT NULL, "
                 "autor TEXT NOT NULL, "
                 "paginas INT NOT NULL, "
                 "ficcao INT, "
                 "nao_ficcao INT, "
                 "romance INT, "
                 "horror INT, "
                 "biografia INT);";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        cerr << mensagemErro << sqlite3_errmsg(bancoDados) << endl;
        sqlite3_free(erroBanco);
    }

    sqlite3_close(bancoDados);
}

/**
 * Edita as informacoes de livro.
 *
 * @param livro
 * @return 0 (sucesso) e 1 (erro)
 */
int editaLivro(struct Livro &livro) {
    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao editar as informaçéos do livro: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "UPDATE livro "
                 "SET nome = '" + livro.nome + "', autor = '" + livro.autor + "', paginas = " +
                 to_string(livro.paginas) + ", ficcao = " + to_string(livro.ficcao) + ", nao_ficcao = " +
                 to_string(livro.naoFiccao) + ", romance = " + to_string(livro.romance) + ", horror = " +
                 to_string(livro.horror) + ", biografia = " + to_string(livro.biografia) + " WHERE id = " +
                 to_string(livro.id) + ";";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_free(erroBanco);
        sqlite3_close(bancoDados);

        return 1;
    }

    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Insere as informacoes de livro.
 *
 * @param livro
 * @return 0 (sucesso) e 1 (erro)
 */
int insereLivro(struct Livro &livro) {
    livro.id = geraIDLivro();

    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao inserir livro: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "INSERT INTO livro(id, nome, autor, paginas, ficcao, nao_ficcao, romance, horror, biografia) "
                 "VALUES('" + to_string(livro.id) + "', '" + livro.nome + "', '" + livro.autor + "', '" +
                 to_string(livro.paginas) + "', '" + to_string(livro.ficcao) + "', '" + to_string(livro.naoFiccao) +
                 "', '" + to_string(livro.romance) + "', '" + to_string(livro.horror) + "', '" +
                 to_string(livro.biografia) + "');";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_free(erroBanco);
        sqlite3_close(bancoDados);

        return 1;
    }

    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Gera o proximo ID.
 * @param id
 * @return ultimo id (sucesso) e -1 (erro)
 */
int geraIDLivro() {
    int id = 1;
    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao consultar o último ID do livro: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return -1;
    }

    string sql = "SELECT id FROM livro ORDER BY id DESC LIMIT 1;";
    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return -1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE) break;

        if (retorno != SQLITE_ROW) {
            exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return -1;
        }
        id = sqlite3_column_int(stmt, 0);
        id += 1;
    }
    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return id;
}

/**
 * Lista todos os livros.
 *
 * @param livros
 * @return 0 (sucesso) e 1 (erro)
 */
int listaLivros(vector<struct Livro> &livros) {
    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao listar livros: ";
    livros.clear();

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "SELECT *, (SELECT COUNT(id_livro) FROM estante WHERE id_livro = id) AS leitores, "
                 "(SELECT AVG(nota) FROM estante WHERE id_livro = id) AS nota_geral FROM livro;";

    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE) break;

        if (retorno != SQLITE_ROW) {
            exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return 1;
        }

        struct Livro livro;
        livro.id = sqlite3_column_int(stmt, 0);
        livro.nome = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 1));
        livro.autor = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 2));
        livro.paginas = sqlite3_column_int(stmt, 3);
        livro.ficcao = sqlite3_column_int(stmt, 4);
        livro.naoFiccao = sqlite3_column_int(stmt, 5);
        livro.romance = sqlite3_column_int(stmt, 6);
        livro.horror = sqlite3_column_int(stmt, 7);
        livro.biografia = sqlite3_column_int(stmt, 8);
        livro.leitores = sqlite3_column_int(stmt, 9);
        livro.notaGeral = sqlite3_column_double(stmt, 10);
        livros.push_back(livro);
    }

    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return 0;
}

 /**
  * Pesquisa livros por titulo.
  *
  * @param livros
  * @param titulo
  * @return 0 (sucesso) e 1 (erro)
  */
int pesquisaLivros(vector<struct Livro> &livros, string titulo) {
    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao pesquisar livros por titulo: ";
    livros.clear();

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "SELECT *, (SELECT COUNT(id_livro) FROM estante WHERE id_livro = id) AS leitores, "
                 "(SELECT AVG(nota) FROM estante WHERE id_livro = id) AS nota_geral FROM livro WHERE nome like '%" +
                 titulo +"%';";

    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE) break;

        if (retorno != SQLITE_ROW) {
            exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return 1;
        }

        struct Livro livro;
        livro.id = sqlite3_column_int(stmt, 0);
        livro.nome = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 1));
        livro.autor = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 2));
        livro.paginas = sqlite3_column_int(stmt, 3);
        livro.ficcao = sqlite3_column_int(stmt, 4);
        livro.naoFiccao = sqlite3_column_int(stmt, 5);
        livro.romance = sqlite3_column_int(stmt, 6);
        livro.horror = sqlite3_column_int(stmt, 7);
        livro.biografia = sqlite3_column_int(stmt, 8);
        livro.leitores = sqlite3_column_int(stmt, 9);
        livro.notaGeral = sqlite3_column_double(stmt, 10);
        livros.push_back(livro);
    }

    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Retorna os livros a serem recomendados para o usuario logado.
 *
 * @param livros
 * @param usuario
 * @return
 */
int recomendaLivros(vector<struct Livro> &livros, struct Usuario &usuario) {
    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao recomendar livros para usuario: ";
    livros.clear();

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "SELECT *, (SELECT COUNT(id_livro) FROM estante WHERE id_livro = id) AS leitores, "
                 "(SELECT AVG(nota) FROM estante WHERE id_livro = id) AS nota_geral FROM livro ";

    string where = "WHERE ";
    where += (usuario.ficcao) ? "ficcao = 1 OR " : "";
    where += (usuario.naoFiccao) ? "nao_ficcao = 1 OR " : "";
    where += (usuario.romance) ? "romance = 1 OR " : "";
    where += (usuario.horror) ? "horror = 1 OR " : "";
    where += (usuario.biografia) ? "biografia = 1 OR " : "";
    where += "FALSE;";
    sql += where;

    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE) break;

        if (retorno != SQLITE_ROW) {
            exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return 1;
        }

        if (sqlite3_column_type(stmt, 0) != SQLITE_NULL) {
            struct Livro livro;
            livro.id = sqlite3_column_int(stmt, 0);
            livro.nome = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 1));
            livro.autor = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 2));
            livro.paginas = sqlite3_column_int(stmt, 3);
            livro.ficcao = sqlite3_column_int(stmt, 4);
            livro.naoFiccao = sqlite3_column_int(stmt, 5);
            livro.romance = sqlite3_column_int(stmt, 6);
            livro.horror = sqlite3_column_int(stmt, 7);
            livro.biografia = sqlite3_column_int(stmt, 8);
            livro.leitores = sqlite3_column_int(stmt, 9);
            livro.notaGeral = sqlite3_column_double(stmt, 10);
            livros.push_back(livro);
        }
    }

    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Remove o livro pelo id.
 *
 * @param id
 * @return 0 (sucesso) e 1 (erro)
 */
int removeLivro(int id) {
    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao remover livro: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "DELETE FROM livro "
                 "WHERE id = " + to_string(id) + ";";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_free(erroBanco);
        sqlite3_close(bancoDados);

        return 1;
    }

    sqlite3_close(bancoDados);

    return 0;
}

File: funcional/src/LivroDB.hs
module LivroDB where

import           Database.HDBC
import           Database.HDBC.Sqlite3
import           EstanteType

-- Cria uma tabela de livro
criaTabelaLivro :: IO ()
criaTabelaLivro = do
  conexao <- connectSqlite3 "prompskoob.db"
  run
    conexao
    ("CREATE TABLE IF NOT EXISTS livro(id INTEGER PRIMARY KEY NOT NULL, titulo TEXT NOT NULL, autor TEXT NOT NULL, " ++
     "paginas INT NOT NULL DEFAULT (0), ficcao INT DEFAULT (0), nao_ficcao INT DEFAULT (0), romance INT DEFAULT (0), " ++
     "horror INT  DEFAULT (0), biografia INT DEFAULT (0));")
    []
  commit conexao
  disconnect conexao

-- Consulta o livro por id.
consultaLivro :: Int -> IO Livro
consultaLivro codigo = do
  conexao <- connectSqlite3 "prompskoob.db"
  resultado <- quickQuery' conexao "SELECT * FROM livro WHERE id = ? LIMIT 1;" [toSql codigo]
  disconnect conexao
  if null resultado
    then return (obtemLivroNulo)
    else return (preencheLivro (head resultado))

-- Insere um livro no banco.
insereLivro :: Livro -> IO Bool
insereLivro livro = do
  conexao <- connectSqlite3 "prompskoob.db"
  inserido <-
    run
      conexao
      ("INSERT INTO livro(titulo, autor, paginas, ficcao, nao_ficcao, romance, horror, biografia) VALUES " ++
       "(?, ?, ?, ?, ?, ?, ?, ?);")
      [ toSql (titulo livro)
      , toSql (autor livro)
      , toSql (paginas livro)
      , toSql (ficcao livro)
      , toSql (naoFiccao livro)
      , toSql (romance livro)
      , toSql (horror livro)
      , toSql (biografia livro)
      ]
  commit conexao
  disconnect conexao
  return (inserido >= 1)

-- Lista todos os livros cadastrados no banco.
listaLivros :: IO [Livro]
listaLivros = do
  conexao <- connectSqlite3 "prompskoob.db"
  resultado <-
    quickQuery'
      conexao
      ("SELECT *, COALESCE((SELECT COUNT(id_livro) FROM estante WHERE id_livro = id), 0.0) AS leitores, " ++
       "COALESCE((SELECT AVG(nota) FROM estante WHERE id_livro = id), 0.0) AS nota_geral FROM livro;")
      []
  disconnect conexao
  if null resultado
    then return []
    else return (map preencheLivros resultado)

-- Remove um livro da base de dados.
removeLivro :: Int -> IO Bool
removeLivro codigo = do
  conexao <- connectSqlite3 "prompskoob.db"
  removido <- run conexao "DELETE FROM livro WHERE id = ?;" [toSql codigo]
  commit conexao
  disconnect conexao
  return (removido >= 1)

recomendaLivros :: Usuario -> IO [Livro]
recomendaLivros usuario = do
  conexao <- connectSqlite3 "prompskoob.db"
  resultado <-
    quickQuery'
      conexao
      ("SELECT *, COALESCE((SELECT COUNT(id_livro) FROM estante WHERE id_livro = id), 0) AS leitores, COALESCE((SELECT AVG(nota) FROM " ++
       "estante WHERE id_livro = id), 0.0) AS nota_geral FROM livro WHERE " ++
       whereFiccao ++ whereNaoFiccao ++ whereRomance ++ whereHorror ++ whereBiografia ++ whereCoringa)
      []
  disconnect conexao
  if null resultado
    then return []
    else return (map preencheLivros resultado)
  where
    whereFiccao =
      if gostaFiccao usuario == 1
        then "ficcao = 1 OR "
        else ""
    whereNaoFiccao =
      if gostaNaoFiccao usuario == 1
        then "nao_ficcao = 1 OR "
        else ""
    whereRomance =
      if gostaRomance usuario == 1
        then "romance = 1 OR "
        else ""
    whereHorror =
      if gostaHorror usuario == 1
        then "horror = 1 OR "
        else ""
    whereBiografia =
      if gostaBiografia usuario == 1
        then "biografia = 1 OR "
        else ""
    whereCoringa = " FALSE;"

editaLivro :: Livro -> IO Livro
editaLivro livro = do
  conexao <- connectSqlite3 "prompskoob.db"
  run
    conexao
    ("UPDATE livro SET titulo = ?, autor = ?, paginas = ?, ficcao = ?, nao_ficcao = ?, romance = ?, horror = ?, " ++
     "biografia = ? WHERE id = ?;")
    [ toSql (titulo livro)
    , toSql (autor livro)
    , toSql (paginas livro)
    , toSql (ficcao livro)
    , toSql (naoFiccao livro)
    , toSql (romance livro)
    , toSql (horror livro)
    , toSql (biografia livro)
    , toSql (idLivro livro)
    ]
  commit conexao
  disconnect conexao
  livroEditado <- consultaLivro (idLivro livro)
  return (livroEditado)

pesquisaLivros :: String -> IO [Livro]
pesquisaLivros titulo = do
  conexao <- connectSqlite3 "prompskoob.db"
  resultado <-
    quickQuery'
      conexao
      ("SELECT *, COALESCE((SELECT COUNT(id_livro) FROM estante WHERE id_livro = id),0) AS leitores, COALESCE((SELECT AVG(nota) FROM " ++
       "estante WHERE id_livro = id),0) AS nota_geral FROM livro WHERE titulo like '%" ++ titulo ++ "%';")
      []
  disconnect conexao
  if null resultado
    then return []
    else return (map preencheLivros resultado)

preencheLivro :: [SqlValue] -> Livro
preencheLivro [sqlId, sqlNome, sqlAutor, sqlPaginas, sqlFiccao, sqlNaoFiccao, sqlRomance, sqlHorror, sqlBiografia] =
  Livro
    { idLivro = codigoSql
    , titulo = nomeSql
    , autor = autorSql
    , paginas = paginasSql
    , ficcao = ficcaoSql
    , naoFiccao = naoFiccaoSql
    , romance = romanceSql
    , horror = horrorSql
    , biografia = biografiaSql
    , leitores = 0
    , notaGeral = 0.0
    }
  where
    codigoSql = fromSql sqlId :: Int
    nomeSql = fromSql sqlNome :: String
    autorSql = fromSql sqlAutor :: String
    paginasSql = fromSql sqlPaginas :: Int
    ficcaoSql = fromSql sqlFiccao :: Int
    naoFiccaoSql = fromSql sqlNaoFiccao :: Int
    romanceSql = fromSql sqlRomance :: Int
    horrorSql = fromSql sqlHorror :: Int
    biografiaSql = fromSql sqlBiografia :: Int

preencheLivros :: [SqlValue] -> Livro
preencheLivros [sqlId, sqlNome, sqlAutor, sqlPaginas, sqlFiccao, sqlNaoFiccao, sqlRomance, sqlHorror, sqlBiografia, sqlLeitores, sqlNotaGeral] =
  Livro
    { idLivro = codigoSql
    , titulo = nomeSql
    , autor = autorSql
    , paginas = paginasSql
    , ficcao = ficcaoSql
    , naoFiccao = naoFiccaoSql
    , romance = romanceSql
    , horror = horrorSql
    , biografia = biografiaSql
    , leitores = leitoresSql
    , notaGeral = notaGeralSql
    }
  where
    codigoSql = fromSql sqlId :: Int
    nomeSql = fromSql sqlNome :: String
    autorSql = fromSql sqlAutor :: String
    paginasSql = fromSql sqlPaginas :: Int
    ficcaoSql = fromSql sqlFiccao :: Int
    naoFiccaoSql = fromSql sqlNaoFiccao :: Int
    romanceSql = fromSql sqlRomance :: Int
    horrorSql = fromSql sqlHorror :: Int
    biografiaSql = fromSql sqlBiografia :: Int
    leitoresSql = fromSql sqlLeitores :: Int
    notaGeralSql = fromSql sqlNotaGeral :: Double

obtemLivroNulo :: Livro
obtemLivroNulo = Livro 0 "NULL" "NULL" 0 0 0 0 0 0 0 0.0

isLivroNulo :: Livro -> Bool
isLivroNulo livro
  | (idLivro livro == 0) && (titulo livro == "NULL") = True
  | otherwise = False
